src/ImageRunModal.jsx

```diff
@@ -27,13 +28,107 @@ const units = {
     },
 };
 
+class PublishPortsForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            publish: [{ containerPort: null, hostPort: null, protocol: 'TCP' }],
+        };
+        this.removePort = this.removePort.bind(this);
+        this.addPort = this.addPort.bind(this);
+        this.onPortChange = this.onPortChange.bind(this);
+    }
+
+    removePort(idx, field, value) {
+        this.setState(state => {
+            let ports = state.publish.concat();
+            ports.splice(idx, 1);
+            if (ports.length === 0)
+                ports.push({ containerPort: null, hostPort: null, protocol: 'TCP' });
+            return { publish: ports };
+        }, () => this.props.onChange(this.state.publish.concat()));
+    }
+
+    addPort() {
+        this.setState(state => {
+            return { publish: [...state.publish, { containerPort: null, hostPort: null, protocol: 'TCP' }] };
+        }, () => this.props.onChange(this.state.publish.concat()));
+    }
+
+    onPortChange(idx, field, value) {
+        this.setState(state => {
+            let ports = state.publish.concat();
+            ports[idx][field] = value || null;
+            return { publish: ports };
+        }, () => this.props.onChange(this.state.publish.concat()));
+    }
+
+    render () {
+        const { id } = this.props;
+        const dialogValues = this.state;
+        return (
+            <React.Fragment>
+                {
+                    dialogValues.publish.map((port, idx) =>
+                        (
+                            <div className='publish-port-form ct-form-layout' key={'published-port-' + idx}>
+                                <div role='group' className='ct-form-layout-split'>
+                                    <input className='form-control'
+                                           id={ (idx === 0 && id) || '' }
```

Comment:
Let's please not use an empty ID, that smells odd. Dropping the fallback should use `undefined`, if React complains let's use `null`?